File: 백준/Gold/12919. A와 B 2/A와 B 2.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const S = input[0].split("")
const T = input[1].split("")

// 맨뒤가 A이고
//  맨 앞이 A
//  -> pop
//  맨 앞이 B
//  -> 뒤집고 pop
//  -> pop

// 맨뒤가 B이고
//  맨 앞이 B
//  -> 뒤집고 pop
const solution = (T, S) => {
  const answer = []
  const getT = (T, S) => {

    if (T.length > S.length) {
      if (T.at(-1) == "B") {
        if (T[0] == "B") {
          const tmpT = [...T]
          tmpT.reverse()
          getT(tmpT.slice(0, -1), S)
        }
      } else {
        if (T[0] == "A") {
          getT(T.slice(0, -1), S)
        } else {
          const tmpT = [...T]

          getT(tmpT.slice(0, -1), S)
          tmpT.reverse()
          getT(tmpT.slice(0, -1), S)
        }
      }
    }else{
        answer.push(T.join("") == S.join(""))
    }
  }
  getT(T, S)
  return answer.some((val) => val === true)
}

console.log(Number(solution(T, S)))

File: 백준/Gold/12919. A와 B 2/A와 B 2.py
import sys
input = sys.stdin.readline

s = list(input().rstrip())
t = list(input().rstrip())
answer = 0
def dfs(t,s):
  global answer
  if(len(t) == len(s)):
    if(''.join(t) == ''.join(s)):
      answer = 1
      return
  else:
    if(t[-1] == 'A'):
      dfs(t[0:-1],s)
    if(t[0] == 'B'):
      tmp = t[::-1]
      dfs(tmp[0:-1],s)

dfs(t,s)
print(answer)
